### PR TITLE
Add simulated SITL RGBLed using sf windowing

### PR DESCRIPTION
**Pull request by peterbarker**
.... kind of like the OSD but... somewhat simpler...

### CODE REVIEW

**Comment by WickedShell** [comment]
SITL manipulates the board led GPIO pins already, why not just fix the defines and hook a SITL GPIO display up?

**Comment by peterbarker** [comment]
On Wed, 15 May 2019, WickedShell wrote:

> SITL manipulates the board led GPIO pins already, why not just fix the defines and hook a SITL GPIO display up?

Not board LEDs?

File: libraries/AP_Notify/SITL_SFML_LED.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <AP_HAL/AP_HAL.h>
+
+#include <stdio.h>
+#include <unistd.h>
+
+#ifdef WITH_SITL_RGBLED
```
Comment (OXINARF) [inline libraries/AP_Notify/SITL_SFML_LED.cpp]: Nitpicking: putting this first prevents recompilation of this file if something changes in AP_HAL.h (or the C headers).

File: libraries/AP_Notify/SITL_SFML_LED.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+// trampoline for update thread
+void *SITL_SFML_LED::update_thread_start(void *obj)
+{
+    ((SITL_SFML_LED *)obj)->update_thread();
+    return nullptr;
+}
+
+bool SITL_SFML_LED::hw_init()
+{
+    pthread_create(&thread, NULL, update_thread_start, this);
```
Comment (OXINARF) [inline libraries/AP_Notify/SITL_SFML_LED.cpp]: Is there a reason why this doesn't use the HAL API for creating a thread?

File: libraries/AP_Notify/SITL_SFML_LED.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    sf::RenderWindow *w = new sf::RenderWindow(sf::VideoMode(width, height), "LED");
+    if (!w) {
+        AP_HAL::panic("Unable to create RGBLed window");
+    }
+
+    while (w->isOpen()) {
+        sf::Event event;
+        while (w->pollEvent(event)) {
+            if (event.type == sf::Event::Closed) {
+                w->close();
```
Comment (OXINARF) [inline libraries/AP_Notify/SITL_SFML_LED.cpp]: Should it return here?

**Comment by tridge** [comment]
merged after discussion on dev call

**Comment by peterbarker** [comment]
On Mon, 20 May 2019, Francisco Ferreira wrote:

> I don't understand the mutex commits, where are the different threads?

The windowing stuff *really* doesn't like multiple threads.  The windowing 
is done from the pthread_create'd calls.

> +#include <AP_HAL/AP_HAL.h>
> +
> +#include <stdio.h>
> +#include <unistd.h>
> +
> +#ifdef WITH_SITL_RGBLED
> 
> Nitpicking: putting this first prevents recompilation of this file if something changes in AP_HAL.h (or the C headers).

Fair call.  Let's make a habit of that :-)

> +bool SITL_SFML_LED::hw_init()
> +{
> +    pthread_create(&thread, NULL, update_thread_start, this);
> 
> Is there a reason why this doesn't use the HAL API for creating a thread?

I swiped it from the OSD code; I can't think there's a good reason, but at 
least these look the same.

> +    while (w->isOpen()) {
> +        sf::Event event;
> +        while (w->pollEvent(event)) {
> +            if (event.type == sf::Event::Closed) {
> +                w->close();
> 
> Should it return here?

Yeah, I debated that.  I decided to stick to existing behaviour copied 
from OSD (IIRC...)

**Comment by OXINARF** [comment]
> The windowing stuff *really* doesn't like multiple threads.  The windowing
is done from the pthread_create'd calls.

I've finally understood what I had missed, the OSD is also using the AP_Notify mutex! That should really be in the HAL instead (or you could go ahead with AP_Telemetry so that AP_Notify and AP_OSD can be types of telemetry 😛).

**Comment by peterbarker** [comment]
On Tue, 21 May 2019, Francisco Ferreira wrote:

> I've finally understood what I had missed, the OSD is also using the AP_Notify mutex! That should really be in the HAL instead (or you could go ahead with

I wasn't happy with where I put it and said that on the call.  Nobody came 
up with anything better - and I'm not sure HAL is the correct answer....

> AP_Telemetry so that AP_Notify and AP_OSD can be types of telemetry 😛).

Well, that's the GCS class - it's unlikely we'll be renaming it in the 
near future.  Moving those two classes to have their instances in GCS 
(like frsky) is a distinct possibility.

**Pull request merged**